### ISSUE TITLE
Packing in DPOTrainer

### ISSUE THREAD

**Issue by zhc7** [huggingface/trl#2469]
### Feature request

packing can be supported in dpo trainer.

### Motivation

sequences with high variation in length padded together will waste a lot of resources. issue #1274 mentioned this, but i don't think the conclusion is correct. packing won't conflict with pairwise datasets. you just need to unpack the sequence after forwarding. you can easily identify the start and the end of each sequence by the position of 0 in `position_ids`.

### Your contribution

actually i already have a version of dpo trainer that can deal with packing:

```python
    def _forward_one(self, model: nn.Module, batch: dict, name: str):
        batch = dict(batch)
        targets = batch.pop("targets")
        loss_masks = batch.pop("loss_masks").to(torch.bool)
        out = model(**batch)
        logits = out["logits"]
        targets = targets.clone()
        targets[~loss_masks] = 0  # dummy token; we'll ignore the losses on these tokens later
        per_token_logps = torch.gather(logits.log_softmax(-1), dim=2, index=targets.unsqueeze(2)).squeeze(2)
        per_token_logps[~loss_masks] = 0
        # unpack logp
        assert batch["position_ids"].shape[0] == 1
        starts = (batch["position_ids"] == 0).nonzero()[:, 1]
        seqs = []
        for i in range(len(starts) - 1):
            seqs.append(per_token_logps[0, starts[i]: starts[i + 1]].sum())
        seqs.append(per_token_logps[0, starts[-1]:].sum())
        all_logps = torch.stack(seqs)
        return {
            f"{name}_logps": all_logps,
            f"mean_{name}_logits": logits[loss_masks].mean(),
        }

    def concatenated_forward(self, model: nn.Module, batch: dict):
        ret = {}
        for name in ("chosen", "rejected"):
            ret = {**ret, **self._forward_one(model, batch[name], name)}
        return ret
```

this does not directly fit into the complex logic of dpo trainer, but the idea is that it is possible to do packing.

**Comment by qgallouedec** [comment]
What you're describing sounds closer to _padding-free_ than packing. We have a (currently draft) PR for this: https://github.com/huggingface/trl/pull/2437

---

At this point I'm not even sure that packing for DPO makes sense. How to ensure that you've as many chosen than rejected? How to ensure they match? How to handle partial sequences?

**Comment by zhc7** [comment]
Hi, thank you for your response. I looked into the link you provided. I think we are talking about the same thing. I used the word "packing" from https://huggingface.co/blog/packing-with-FA2. The "packing" here actually means concatenating a fixed batch size of samples into one sequence, and use `position_ids` to mark the boundaries, rather than packing to a fixed length. So there won't be the problems you mentioned. I've also briefly read https://huggingface.co/blog/mayank-mishra/padding-free-transformer this blog, I think the ideas are the same. But I'm not sure how the latter is implemented. Maybe they are the same thing just with different names:)

I breifly went through the pr, I see it is trying to add `position_ids` in the whole process, so I guess we are talking about the same thing.

**Comment by qgallouedec** [comment]
new draft PR: https://github.com/huggingface/trl/pull/5141

**Comment by albertvillanova** [comment]
Thanks @zhc7. 

The padding-free feature was accepted and shipped in #2520 (merged 2025-01-08), then disabled by the DPO refactor #3906. On main, DPOConfig(padding_free=True) silently no-ops: it warns and falls back to padded https://github.com/huggingface/trl/blob/2396dfe5d2be7b18c0b615d80957d64ecdeb7cc0/trl/trainer/dpo_trainer.py#L717-L724 so a documented public config field is dead for every DPO user.

Restoration is @qgallouedec draft PR #5141 (last touched 2026-07-29).

**Comment by behroozazarkhalili** [comment]
Opened #6972 for the documentation half of this.

It does not restore the feature: @qgallouedec said on #5141 that he would redo padding-free DPO from a fresh branch, so that work stays his. #6972 only makes the current state visible, in the `DPOConfig.padding_free` docstring, in the field's `help` text, and in the DPO tab of the padding-free docs page, so a user reading any of the three learns that the flag falls back to standard padding today.

One detail worth recording here, since the runtime warning does not mention it. Measured against `trl-internal-testing/tiny-Qwen3ForCausalLM`:

```
requested=True   config.padding_free=True   trainer.padding_free=False
requested=False  config.padding_free=False  trainer.padding_free=False
```

`dpo_trainer.py:717` copies `args.padding_free` into an instance attribute and `:724` overrides that copy, not the config. So `training_args.padding_free` keeps reading `True` for the whole run, and anything that serializes the config records a feature that never ran.
